extras/gizmo/gizmo.js

```diff
@@ -0,0 +1,351 @@
+        this._coordSpace = value ?? 'world';
+        this._updateRotation();
+        this.fire('coordSpace:set', this._coordSpace);
+    }
+
+    get coordSpace() {
+        return this._coordSpace;
+    }
+
+    set size(value) {
+        this._size = value;
+        this._updateScale();
+        this.fire('size:set', this._size);
+    }
+
+    get size() {
+        return this._size;
+    }
+
+    _getProjFrustumWidth() {
+        const gizmoPos = this.gizmo.getPosition();
+        const cameraPos = this.camera.getPosition();
+        const dist = tmpV1.copy(gizmoPos).sub(cameraPos).dot(this.camera.forward);
+        return dist * Math.tan(this.camera.camera.fov * math.DEG_TO_RAD / 2);
+    }
+
+    _createLayer() {
+        const layerMap = this.app.scene.layers.layerIdMap;
+        if (layerMap.has(GIZMO_LAYER_ID)) {
+            this.layer = layerMap.get(GIZMO_LAYER_ID);
+        } else {
+            this.layer = new Layer({
+                id: GIZMO_LAYER_ID,
+                name: 'Gizmo',
+                clearDepthBuffer: true,
+                opaqueSortMode: SORTMODE_NONE,
+                transparentSortMode: SORTMODE_NONE
+            });
+            this.app.scene.layers.push(this.layer);
+        }
+        if (this.camera.camera.layers.indexOf(this.layer.id) === -1) {
+            this.camera.camera.layers = this.camera.camera.layers.concat(this.layer.id);
+        }
+    }
+
+    _createGizmo() {
+        this.gizmo = new Entity('gizmo');
+        this.app.root.addChild(this.gizmo);
+        this.gizmo.enabled = false;
+    }
+
+    _updatePosition() {
+        tmpV1.set(0, 0, 0);
+        for (let i = 0; i < this.nodes.length; i++) {
+            const node = this.nodes[i];
+            tmpV1.add(node.getPosition());
+        }
+        tmpV1.scale(1.0 / this.nodes.length);
+        this.gizmo.setPosition(tmpV1);
+
+        this.fire('position:set', tmpV1);
+    }
+
+    _updateRotation() {
+        if (this._coordSpace === 'local') {
+            tmpV1.set(0, 0, 0);
+            for (let i = 0; i < this.nodes.length; i++) {
+                const node = this.nodes[i];
+                tmpV1.add(node.getEulerAngles());
+            }
+            tmpV1.scale(1.0 / this.nodes.length);
+            this.gizmo.setEulerAngles(tmpV1);
+        } else {
+            tmpV1.set(0, 0, 0);
+            this.gizmo.setEulerAngles(tmpV1);
+        }
+
+        this.fire('eulerAngles:set', tmpV1);
+    }
+
+    _updateScale() {
+        let scale = 1;
+        if (this.camera.camera.projection === PROJECTION_PERSPECTIVE) {
+            scale = this._getProjFrustumWidth() * PERS_SCALE_RATIO;
+        } else {
+            scale = this.camera.camera.orthoHeight * ORTHO_SCALE_RATIO;
+        }
+        scale = Math.max(scale * this._size, MIN_GIZMO_SCALE);
+        tmpV1.set(scale, scale, scale);
+        this.gizmo.setLocalScale(tmpV1);
+
+        this.fire('scale:set', tmpV1);
+
+    }
+
+    _getSelection(x, y) {
```

Comment:
This feels rather expensive. Maybe the gizmo should not use entities / render components at all, and only create meshInstances and insert those to the layer directly.
We should also not call getPositions / getIndices each time, as those get extracted from storage in the vertex buffer. We should keen our collision data stored directly in a format that is convenient for ray to triangle intersection. I'd add a Triangle class to the engine (constructed from 3 points). And it would have intersect function with the ray. Then this code would store the array of triangles directly.

Reply:
I agree that the engine would benefit from a `Triangle` class that can handle a `Ray` instersection. But I kinda like the idea that this initial implementation is self-contained. Could this be a follow up PR?